### PR TITLE
Recycle buddy java maria

### PR DESCRIPTION
**Pull request by MariaSS88**
I created initial draft for Java part of Back End. 
It uses Maven, Spring Boot and Jackson.

JSON sample structure for a request
{
    "id" : "1",
    "base64" : "abcdefg"
}

It currently returns only "Hello World" as a response for a POST request to /images/v1/recognize

### CODE REVIEW

File: src/main/java/com/recycle/buddy/model/Image.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.recycle.buddy.model;
+
+public class Image {
+
+    private String id;
```
Comment (KAardal) [inline src/main/java/com/recycle/buddy/model/Image.java]: Could these fields be made final? Would there ever be a case where we would need to change their value?

File: src/main/java/com/recycle/buddy/model/Image.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    private String base64;
+
+    public String getId() {
```
Comment (KAardal) [inline src/main/java/com/recycle/buddy/model/Image.java]: If we decide to make the the class immutable, eliminate the setter methods. Either way, I feel that assigning values to the data fields can be done at object creation, so a two arg constructor should probably be provided.

File: src/main/java/com/recycle/buddy/model/Image.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package com.recycle.buddy.model;
+
+public class Image {
+
+    private String id;
+    private String base64;
```
Comment (KAardal) [inline src/main/java/com/recycle/buddy/model/Image.java]: Could these fields be made final? Would there ever be a case where we would need to change their value? I.e can this class be made immutable?

File: src/main/java/com/recycle/buddy/service/ImageRecognitionService.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+
+@Service
+public class ImageRecognitionService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ImageRecognitionService.class);
+
+    public String recognize(final String image) {
```
Comment (KAardal) [inline src/main/java/com/recycle/buddy/service/ImageRecognitionService.java]: Are we sure that this parameter will be of type String?

File: src/main/java/com/recycle/buddy/model/Image.java (hunk 2 above)
Comment (MariaSS88) [reply src/main/java/com/recycle/buddy/model/Image.java]: Jackson framework that binds JSON data with object  requires JavaBean style class.
 I got this message, when tried to use a constructor:
 "Cannot construct instance of `com.recycle.buddy.model.Image` (no Creators, like default construct, exist): cannot deserialize from Object value (no delegate- or property-based Creator)"

File: src/main/java/com/recycle/buddy/model/Image.java (hunk 3 above)
Comment (MariaSS88) [reply src/main/java/com/recycle/buddy/model/Image.java]: As I use Jackson and Spring boot , I have to keep this model and use setters, so fields can not be final.

File: src/main/java/com/recycle/buddy/service/ImageRecognitionService.java (hunk 4 above)
Comment (MariaSS88) [reply src/main/java/com/recycle/buddy/service/ImageRecognitionService.java]: No, we are not sure. I am still waiting an answer from Frond end.

File: src/main/java/com/recycle/buddy/service/ImageRecognitionService.java [hunk 5]
```diff
@@ -1,18 +1,68 @@
...
+            ExamplePayload payload = ExamplePayload.newBuilder().setImage(image).build();
+            Map<String, String> params = new HashMap<>();
+
+            PredictResponse response = predictionServiceClient.predict(name, payload, params);
+
+            System.out.println(response.toString());
+            return response;
+
+        }catch (Throwable t){
+          System.out.println(t);
```
[the remark this answers is not in the document]
Comment (MariaSS88) [reply src/main/java/com/recycle/buddy/service/ImageRecognitionService.java]: Yes, work is in progress. I am not too comfortable with exceptions, so any advises are welcome. I will change it.

File: src/main/java/com/recycle/buddy/service/ImageRecognitionService.java [hunk 6]
```diff
@@ -1,18 +1,68 @@
...
+    //private static final String GCP_LOCATION = "us-central1";
+    //private static final String AUTOML_MODEL_ID = "ICN1265489133416904487";
+
+    public String recognize(final String imageBase64) {
+        LOG.info(imageBase64);
+        PredictResponse response = PredictResponse.getDefaultInstance();
+        try {
+            response = predictResponse(GCP_PROJECT_ID,GCP_LOCATION,AUTOML_MODEL_ID, imageBase64);
+        } catch (Throwable t) {
+            System.out.println(t);
```
[the remark this answers is not in the document]
Comment (MariaSS88) [reply src/main/java/com/recycle/buddy/service/ImageRecognitionService.java]: Made changes

File: src/main/java/com/recycle/buddy/controller/RecycleBuddyController.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    private final ImageRecognitionService imageRecognitionService;
+
+    @Autowired
+    public RecycleBuddyController(final ImageRecognitionService imageRecognitionService) {
+        this.imageRecognitionService = imageRecognitionService;
+    }
+
+    @RequestMapping(value = "/recognize", method = RequestMethod.POST)
+    public String recognize(@RequestBody final Image image) {
```
Comment (AndyCErnst) [inline src/main/java/com/recycle/buddy/controller/RecycleBuddyController.java]: Didn't realize you were waiting on the front end for the API on this. I haven't talked with anyone else but I would argue for returning json in the form of something like
```
{ 
  "data":  { 
     "answers": ["Tin Can", "Aluminum Can" ],
     "minimumConfidence": 0.7 
  },
  "error": "error message"
 }
```
Only include the error field if there is an error. The answers can array can be 0, 1, or multiple entries.
You can do this in a future PR. Let's get this one in

Comment (MariaSS88) [reply src/main/java/com/recycle/buddy/controller/RecycleBuddyController.java]: I was thinking about the similar structure .

**Pull request merged**